src/main/java/org/elasticsearch/search/highlight/HighlightPhase.java

```diff
@@ -116,9 +119,26 @@ public void hitExecute(SearchContext context, HitContext hitContext) {
 
                 Query highlightQuery = field.fieldOptions().highlightQuery() == null ? context.parsedQuery().query() : field.fieldOptions().highlightQuery();
                 HighlighterContext highlighterContext = new HighlighterContext(fieldName, field, fieldMapper, context, hitContext, highlightQuery);
-                HighlightField highlightField = highlighter.highlight(highlighterContext);
-                if (highlightField != null) {
-                    highlightFields.put(highlightField.name(), highlightField);
+
+                if (highlighter.cannotHighlight(fieldMapper) && fieldNameContainsWildcards) {
```

Comment:
thinking out loud, maybe I am getting confused, but in order for a field to get highlighted, doesn't it need to be stored too or we need to have the _source at least? but metadata fields, which match `*` are not part of the `_source` hence they need to be stored or excluded from highlighting by definition. I have the feeling we should do something more to address that...